File: coding_skill_cpp/27_driver_interface.cpp
/* C++ Driver Interface */
#include <iostream>
using namespace std;

class IDriver {
public:
    virtual ~IDriver() = default;
    virtual int open() = 0;
    virtual int close() = 0;
    virtual int read(uint8_t* buf, size_t len) = 0;
    virtual int write(const uint8_t* buf, size_t len) = 0;
};

class UARTDriver : public IDriver {
public:
    int open() override { cout << "[UART] Open" << endl; return 0; }
    int close() override { cout << "[UART] Close" << endl; return 0; }
    int read(uint8_t* buf, size_t len) override { return len; }
    int write(const uint8_t* buf, size_t len) override { return len; }
};

int main() {
    cout << "=== C++ Driver Interface ===" << endl;
    UARTDriver uart;
    uart.open();
    uart.close();
    return 0;
}

File: coding_skill_py/27_driver_interface.py
"""Python Driver Interface - Protocol"""
from typing import Protocol

class IDriver(Protocol):
    def open(self) -> int: ...
    def close(self) -> int: ...
    def read(self, buf: bytearray, length: int) -> int: ...
    def write(self, buf: bytes, length: int) -> int: ...

class UARTDriver:
    def open(self) -> int:
        print("[UART] Open")
        return 0
    
    def close(self) -> int:
        print("[UART] Close")
        return 0
    
    def read(self, buf: bytearray, length: int) -> int:
        return length
    
    def write(self, buf: bytes, length: int) -> int:
        return length

def main():
    print("=== Python Driver Interface ===")
    uart = UARTDriver()
    uart.open()
    uart.close()

if __name__ == "__main__":
    main()
